url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/Button.java

```diff
@@ -0,0 +1,159 @@
+ * 
+ *     sqUIrt is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Lesser General Public License for more details.
+ * 
+ *     You should have received a copy of the GNU Lesser General Public License
+ *     along with sqUIrt.  If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+package nui.squirt.component;
+
+import java.awt.Color;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import nui.squirt.Actionable;
+import nui.squirt.ControlPoint;
+import nui.squirt.event.ActionEvent;
+import nui.squirt.listener.ActionListener;
+import processing.core.PApplet;
+import processing.core.PVector;
+
+
+public class Button extends Rectangle implements Actionable {
+	
+	private static final Color PRESSED_COLOR = Color.GRAY;
+	private static final Color NOT_PRESSED_COLOR = Color.LIGHT_GRAY;
+	private static final Color BORDER_COLOR = Color.BLACK;
+	private static final float BORDER_WIDTH = 4;
+
+	private Label label;
+
+	private boolean pressed;
+
+	private Collection<ActionListener> listeners = new ArrayList<ActionListener>();
+	
+	private ControlPoint current = null;
+	
+	public Button(float x, float y, String text) {
+		super(x, y, 0, 0);
+		this.label = new Label(0, 0, text);
+	}
+
+	public Label getLabel() {
+		return label;
+	}
+	
+	public void setText(String text) {
+		getLabel().setText(text);
+	}
+
+	public boolean isPressed() {
+		return pressed;
+	}
+	
+	public void press() {
+		this.pressed = true;
+	}
+	
+	public void release() {
+		this.pressed = false;
+		fireAction(new ActionEvent(this));
+	}
+
+	public void addActionListener(ActionListener a) {
+		listeners.add(a);
+	}
+
+	public void removerActionListener(ActionListener a) {
+		listeners.remove(a);
+	}
+
+	public void fireAction(ActionEvent e) {
+		for (ActionListener l: listeners) {
+			l.actionPerformed(e);
+		}
+	}
+
+	public String getAction() {
+		return getLabel().getText();
+	}
+
+	public void setAction(String action) {
+		setText(action);
+	}
+	
+	@Override
+	public void update() {
+		super.update();
+		
+		setFillColor(isPressed() ? PRESSED_COLOR : NOT_PRESSED_COLOR);
+		setStrokeColor(BORDER_COLOR);
+		setStrokeWeight(BORDER_WIDTH);
+		
+		setWidth((float) (label.getWidth() * 1.4));
+		setHeight((float) (label.getHeight() * 1.6));
+	}
+	
+	@Override
+	public void render(PApplet p) {
+		super.render(p);
+		
+		getLabel().preRender(p);
+		getLabel().render(p);
+		getLabel().postRender(p);
+	}
+	
+	@Override
+	public boolean canAcceptMoreControlPoints() {
+		return current == null;
+	}
+	
+	@Override
+	public void controlPointCreated(ControlPoint cp) {
+		if (current == null) {
+			// TODO swap code below with call to isUnderPoint()
+			PVector l = transformToLocalSpace(new PVector(cp.getX(), cp.getY()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;cp&#39; (index: 0 | type: ControlPoint). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;controlPointUpdated&#39;, &#39;controlPointDied&#39;. The same parameter type is validated here for example: url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/Circle.java:107. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).